[YUNIKORN-283] Support query params to filter the apps using queue name/namespace for v1/apps endpoint

manirajv06 opened the pull request.

manirajv06 commented:
@yangwwei Should we re-run and see?

wilfred-s commented on pkg/webservice/handlers.go:
```diff
@@ -82,14 +82,23 @@ func getClusterInfo(w http.ResponseWriter, r *http.Request) {
 func getApplicationsInfo(w http.ResponseWriter, r *http.Request) {
 	writeHeaders(w)
 
+	queueName := r.URL.Query().Get("queue")
```
the parameter must be sanity checked to make sure it is a queue:
* must be name <dot> name ...
* each name must comply with QueueNameRegExp (see the configs code for that)
* anything that does not comply give an error or be ignored

wilfred-s commented on pkg/webservice/handlers.go:
```diff
@@ -82,14 +82,23 @@ func getClusterInfo(w http.ResponseWriter, r *http.Request) {
-			appsDao = append(appsDao, appDao)
+			if len(queueName) > 0 {
+				queueNameArr := strings.Split(app.QueueName, ".")
+				appQueueName := queueNameArr[len(queueNameArr)-1]
```
If the app queue name is just the leaf queue than we cannot do this.
The query could be set to `root.test` and the apps runs in `root.sub.test` and your filter will return it as `test` is equal.

We should have the full queue path in the app to correctly filter and check all parts of the query parameter.

wilfred-s commented on pkg/webservice/handlers.go:
```diff
@@ -82,14 +82,23 @@ func getClusterInfo(w http.ResponseWriter, r *http.Request) {
 		for _, app := range appList {
 			appDao := getApplicationJSON(app)
-			appsDao = append(appsDao, appDao)
+			if len(queueName) > 0 {
```
Filtering should happen before we get the whole object

wilfred-s commented:
And I forgot: we need a test for this

manirajv06 commented:
@wilfred-s Addressed all comments. Ended up in moving handlers related test classes into new package "tests" under "webservice" package itself to avoid cycle problems raised after started using scheduler/tests package for newly written test cases. For example, using mock scheduler etc in handlers_test.go#TestQueryParamInAppsHandler. And some more changes because of above said package movement.

adamantal commented on pkg/webservice/tests/handlers_test.go:
```diff
@@ -16,20 +16,37 @@
 
 import (
 	"encoding/json"
+	"github.com/apache/incubator-yunikorn-core/pkg/common/resources"
```
The import orders are not good. Could you please run `make lint` locally to see these warnings before submitting the patch?

wilfred-s commented:
I don't think the approach taken to move the tests is correct.

We already have a number of test methods that will allow us to build up objects without the need to start a scheduler. It looks like there are just two missing: `AddPartitionToCluster` and `AddAppToPartition`. When we add those two in the `pkg/cache/cachetestutils.go` we can create the structures we need.

That removes the need to export a number of functions, start the scheduler and keeps the changes far smaller.

The test would init simply with:
```
	partition, err := cache.CreatePartitionInfo(...)
	app := cache.NewApplicationInfo(...)
	cache.AddAppToPartition(partition, app)
	gClusterInfo = cache.NewClusterInfo() 
	cache.AddPartitionToCluster(gClusterInfo, partition)
```

Please have a look at that option.
Make sure that you run `make test` and `make lint` before you push the changes.

manirajv06 commented:
@wilfred-s Initially made an attempt more or less similar to your suggestion and then changed approach (of starting scheduler etc) as I had some issues which didn't worked even after a couple of tries. Glad you stressed and tried again now with good closure. Thanks.

manirajv06 commented:
BTW, had come across intermittent test failures. However, I don't think it is related.

--- FAIL: TestFairnessAllocationForNodes (2.28s)    testing.go:906: race detected during execution of test

wilfred-s commented on pkg/webservice/handlers_test.go:
```diff
@@ -175,3 +177,87 @@ func TestContainerHistory(t *testing.T) {
+	// add a new app
+	appInfo := cache.CreateNewApplicationInfo("app-1", partitionName, "root.default")
+	err := cache.AddAppToPartition(appInfo, defaultPartition)
+	if err != nil {
```
use `assert.NilError`

wilfred-s commented on pkg/webservice/handlers_test.go:
```diff
@@ -175,3 +177,87 @@ func TestContainerHistory(t *testing.T) {
+	req1, err1 := http.NewRequest("GET", "/ws/v1/apps?queue=root.q1", strings.NewReader(""))
+	assert.NilError(t, err1, "App Handler request failed")
+	resp1 := &MockResponseWriter{}
+	var appsDao1 []*dao.ApplicationDAOInfo
```
you can reuse `appsDao`, if the conversion fails an error is returned otherwise the content is updated by the Unmarshal

wilfred-s commented on pkg/webservice/handlers_test.go:
```diff
@@ -175,3 +177,87 @@ func TestContainerHistory(t *testing.T) {
+	assert.Equal(t, len(appsDao), 1)
+
+	// Passing "root.q1" as filter return 0 application as there is no app running in "root.q1" queue
+	req1, err1 := http.NewRequest("GET", "/ws/v1/apps?queue=root.q1", strings.NewReader(""))
```
You can reuse variables: `resp`, `req` and `errReq`, no need to create req1..req4
You can even reuse `err` instead of `errReq` by declaring `var req *http.Request` before line 218

wilfred-s commented:
> BTW, had come across intermittent test failures. However, I don't think it is related.
> 
> --- FAIL: TestFairnessAllocationForNodes (2.28s)    testing.go:906: race detected during execution of test

If you have race detections, please copy the run for that test from start to finish and log a jira for it. I cannot remember ever seeing that test show up in any race conditions.
The test start for is often the line which contains `ServiceContext start all services` the end is where the `---FAIL` is logged.

adamantal commented:
If this ever gets pushed in please make sure to update the [/apps REST API documentation](http://yunikorn.apache.org/docs/next/api/scheduler#applications) that YK now supports query params.

manirajv06 commented on pkg/cache/app_test.go:
```diff
@@ -0,0 +1,34 @@
+ limitations under the License.
+*/
+
+func TestKilledStateTransition1(t *testing.T) {
```
Unintentional addition. Removed the file.

wilfred-s replied:
I think the commit did not pick up the removal

manirajv06 commented:
> > BTW, had come across intermittent test failures. However, I don't think it is related.
> > --- FAIL: TestFairnessAllocationForNodes (2.28s)    testing.go:906: race detected during execution of test
> 
> If you have race detections, please copy the run for that test from start to finish and log a jira for it. I cannot remember ever seeing that test show up in any race conditions.
> The test start for is often the line which contains `ServiceContext start all services` the end is where the `---FAIL` is logged.

Ok, sure. Thanks

manirajv06 commented:
@wilfred-s Addressed all review comments. Can you please take a look?

manirajv06 commented:
Updated. Thanks.

wilfred-s commented on pkg/webservice/handlers.go:
```diff
@@ -98,6 +107,20 @@ func getApplicationsInfo(w http.ResponseWriter, r *http.Request) {
+		queueNameArr := strings.Split(queueName, ".")
+		for _, name := range queueNameArr {
+			if !configs.QueueNameRegExp.MatchString(name) {
+				return errors.New("Problem in Queue query parameter parsing as queue param '" + queueName + "' " +
+					"contains invalid queue name '" + name + "'. Queue name must only have alphanumeric characters," +
+					" - or _, and be no longer than 64 characters.")
```
I would recommend that you generate a new error using `fmt.Errorf()`
In the case that there is just a fixed string you can use `errors.New()` when using a complex construct like you do in this case using `fmt.Errorf()` is far more appropriate and improves readability of the code.

wilfred-s commented on pkg/webservice/handlers_test.go:
```diff
@@ -175,3 +177,83 @@ func TestContainerHistory(t *testing.T) {
+
+	configs.MockSchedulerConfigByData([]byte(configDefault))
+	if _, err := cache.SetClusterInfoFromConfigFile(clusterInfo, rmID, policyGroup); err != nil {
+		t.Errorf("Error when load clusterInfo from config %v", err)
```
nit: `t.Fatalf()` gives you this in one go.

wilfred-s commented on pkg/webservice/handlers.go:
```diff
@@ -98,6 +107,20 @@ func getApplicationsInfo(w http.ResponseWriter, r *http.Request) {
+		for _, name := range queueNameArr {
+			if !configs.QueueNameRegExp.MatchString(name) {
+				return fmt.Errorf("problem in queue query parameter parsing as queue param " +
+					"#{queueName} contains invalid queue name #{name}. Queue name must only have " +
```
Not sure what happened here but this does not achieve what you want: use `%s` for the strings see [fmt.Errorf example](https://golang.org/pkg/fmt/#Errorf)

manirajv06 replied:
Has been mis leaded with the way editor showing statements. Corrected now.

wilfred-s commented:
Beside the lint failure the change looks good. Can you have a look: check the space in front of the + signs. When that is done I will approve and commit.

manirajv06 commented:
Fixed. Please take a look at doc changes - https://github.com/apache/incubator-yunikorn-site/pull/12 while merging this.
